src/cli/rustup_mode.rs

```diff
@@ -1159,9 +1166,71 @@ fn component_add(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
         let new_component = Component::new_with_target(component, false)
             .unwrap_or_else(|| Component::new(component.to_string(), target.clone(), true));
 
-        toolchain.add_component(new_component)?;
-    }
+        if m.is_present("allow-downgrade") {
+            let toolchain_date = toolchain
+                .get_manifest()
+                .ok()
+                .and_then(|m| m.map(|m| m.date))
+                .unwrap();
+            let downgrade_date = utc_from_manifest_date(&toolchain_date)
+                .unwrap_or_else(|| panic!("Malformed manifest date: {}", toolchain_date))
+                .pred();
+
+            let release_channel = match toolchain.name() {
+                s if s.starts_with("nightly") => "nightly",
+                s if s.starts_with("stable") => "stable",
+                s if s.starts_with("beta") => "beta",
+                _ => "nightly",
+            };
+            let nightly_date = format!(
+                "{}-{}",
+                &release_channel,
+                &downgrade_date.to_string()[0..10]
+            );
+            let temp_toolchain = cfg.get_toolchain(&nightly_date, false)?;
 
+            let status = if !temp_toolchain.is_custom() {
```

Comment:
This test feels like it belongs at an early bailout rather than so deep into this function.